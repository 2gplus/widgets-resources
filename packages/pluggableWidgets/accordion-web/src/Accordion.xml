<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.web.accordion.Accordion" needsEntityContext="true" offlineCapable="true"
        pluginWidget="true"
        xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Accordion</name>
    <description>Toggle the display of sections of content</description>
    <icon>
        iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAYAAACqaXHeAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAF0SURBVHgB7du9ToRAFAXgw7o2YAt2ttv4EBTuGwCNFnS+kiUkVvAKFjyEja2doRWa1axgYqHDMBnDFsw9Xzlek50T5mcvWe84gGAbCMcAINz270DTNCiKAn3fwyVhGCJNU8Rx/GtceQLqunZu8qO2bVGWpTK+mSp0Vdd1yhj3AFNBVVVYsyzLZv/OYxDCcQ+ApafXTzw8f6A7LPsV4tL3cLfb4ubqbLZOd0/RnfMm1k/A48vykx+99cch2IOxTndP0Z3zJtYBvJs/479dnHvGmqmz/EcQBLBlHcD99fb7cV1aMEz+dmdekXmeI4oi9f+HySdJAlvWe8B+WKN7wzo9pXGN267zOTwGIRzvAaYC01167bgEIBz3AFMB+wGOYwAQjv0AWGI/gP0A9gPYD3AJA4Bw7AdAOAYA4dgPgHAMAMIxAAjHACAcA4BwDADCMQAIxwAgHHuCEE4JwPd9uGrqlZoSgO7d29rp3h16/O2wcAwAwn0BL4eSOakvH2IAAAAASUVORK5CYII=
    </icon>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="groups" type="object" isList="true" required="true">
                    <caption>Groups</caption>
                    <description/>
                    <properties>
                        <propertyGroup caption="General">
                            <property key="headerRenderMode" type="enumeration" defaultValue="text">
                                <caption>Header</caption>
                                <description/>
                                <enumerationValues>
                                    <enumerationValue key="text">Text</enumerationValue>
                                    <enumerationValue key="custom">Custom</enumerationValue>
                                </enumerationValues>
                            </property>
                            <property key="headerText" type="textTemplate">
                                <caption>Text</caption>
                                <description/>
                                <translations>
                                    <translation lang="en_US">Header</translation>
                                    <translation lang="nl_NL">Koptekst</translation>
                                </translations>
                            </property>
                            <property key="headerContent" type="widgets" required="false">
                                <caption>Header</caption>
                                <description/>
                            </property>
                            <property key="content" type="widgets" required="false">
                                <caption>Content</caption>
                                <description/>
                            </property>
                            <property key="visible" type="expression" defaultValue="true">
                                <caption>Visible</caption>
                                <description/>
                                <returnType type="Boolean"/>
                            </property>
                            <property key="dynamicClass" type="expression" required="false">
                                <caption>Dynamic class</caption>
                                <description/>
                                <returnType type="String"/>
                            </property>
                        </propertyGroup>
                    </properties>
                </property>
            </propertyGroup>
            <propertyGroup caption="Behavior">
                <property key="collapsible" type="boolean" defaultValue="true">
                    <caption>Collapsible</caption>
                    <description/>
                </property>
                <property key="collapseBehavior" type="enumeration" defaultValue="singleExpanded">
                    <caption>Collapse behavior</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="singleExpanded">Only one group can be expanded</enumerationValue>
                        <enumerationValue key="multipleExpanded">Multiple groups can be expanded</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="animate" type="boolean" defaultValue="true">
                    <caption>Animate</caption>
                    <description/>
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Visualization">
            <propertyGroup caption="Icon">
                <property key="showIcon" type="enumeration" defaultValue="right">
                    <caption>Show icon</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="right">Right</enumerationValue>
                        <enumerationValue key="left">Left</enumerationValue>
                        <enumerationValue key="no">No</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="icon" type="icon" required="false">
                    <caption>Icon</caption>
                    <description/>
                </property>
                <property key="expandIcon" type="icon" required="false">
                    <caption>Expand icon</caption>
                    <description/>
                </property>
                <property key="collapseIcon" type="icon" required="false">
                    <caption>Collapse icon</caption>
                    <description/>
                </property>
                <property key="animateIcon" type="boolean" defaultValue="true">
                    <caption>Animate icon</caption>
                    <description>Animate the icon when the group is collapsing or expanding</description>
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Advanced">
            <property key="advancedMode" type="boolean" defaultValue="false">
                <caption>Advanced options</caption>
                <description/>
            </property>
        </propertyGroup>
    </properties>
</widget>
