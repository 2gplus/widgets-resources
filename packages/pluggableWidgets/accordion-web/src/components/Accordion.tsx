import { createElement, ReactElement, useMemo } from "react";

import AccordionGroup, { AccordionGroupProps } from "./AccordionGroup";

import { AccordionContainerProps } from "../../typings/AccordionProps";

export interface AccordionProps extends Pick<AccordionContainerProps, "class" | "style" | "tabIndex"> {
    id: string;
    groups: AccordionGroupProps[];
}

export default function Accordion(props: AccordionProps): ReactElement | null {
    const { id, class: classNames, style, tabIndex, groups } = props;

    const renderedGroups = useMemo(() => {
        return groups.map((group, index) => (
            <AccordionGroup key={index} header={group.header} content={group.content} />
        ));
    }, [groups]);

    return (
        <div id={id} className={classNames} style={style} tabIndex={tabIndex}>
            {renderedGroups}
        </div>
    );
}
