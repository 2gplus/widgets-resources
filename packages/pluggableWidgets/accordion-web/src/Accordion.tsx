import { createElement, ReactElement, useMemo } from "react";
import { ValueStatus } from "mendix";

import { Accordion as AccordionComponent, AccordionGroups } from "./components/Accordion";
import { useIconGenerator } from "./utils/iconGenerator";

import { AccordionContainerProps } from "../typings/AccordionProps";

export function Accordion(props: AccordionContainerProps): ReactElement | null {
    const accordionGroups: AccordionGroups | undefined = useMemo(() => {
        if (props.groups.some(group => group.visible.value === undefined || group.headerText.value === undefined)) {
            return undefined;
        }

        return props.groups.map(group => ({
            header: group.headerRenderMode === "text" ? <h3>{group.headerText.value}</h3> : group.headerContent,
            content: group.content,
            visible: group.visible.value!,
            dynamicClassName: group.dynamicClass?.value
        }));
    }, [props.groups]);

    const generateIcon = useIconGenerator(
        props.advancedMode,
        props.animateIcon,
        { data: props.icon?.value, loading: props.icon?.status === ValueStatus.Loading },
        { data: props.expandIcon?.value, loading: props.expandIcon?.status === ValueStatus.Loading },
        { data: props.collapseIcon?.value, loading: props.collapseIcon?.status === ValueStatus.Loading }
    );

    if (!accordionGroups) {
        return null;
    }

    return (
        <AccordionComponent
            id={props.name}
            class={props.class}
            style={props.style}
            tabIndex={props.tabIndex}
            groups={accordionGroups}
            collapsible={props.collapsible}
            animateCollapsing={props.animate}
            singleExpandedGroup={props.collapsible ? props.expandBehavior === "singleExpanded" : undefined}
            generateHeaderIcon={generateIcon}
            showGroupHeaderIcon={props.showIcon}
        />
    );
}
