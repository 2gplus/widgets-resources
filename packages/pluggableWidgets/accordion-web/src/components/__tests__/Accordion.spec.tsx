import { createElement } from "react";
import { mount } from "enzyme";
import { Accordion, AccordionProps } from "../Accordion";

describe("Accordion", () => {
    let defaultProps: AccordionProps;

    function getProps(collapsible: boolean, singleExpandedGroup?: boolean): AccordionProps {
        return {
            id: "id",
            class: "",
            style: { height: "500px" },
            tabIndex: 1,
            groups: [
                { header: "header", content: <span>content</span>, visible: true },
                { header: "header2", content: <span>content2</span>, visible: false }
            ],
            collapsible,
            singleExpandedGroup,
            generateHeaderIcon: jest.fn(),
            showGroupHeaderIcon: "right"
        };
    }

    describe("in collapsible & single expanded group mode", () => {
        beforeEach(() => {
            defaultProps = getProps(true, true);
        });

        it("renders correctly", () => {
            const accordion = mount(<Accordion {...defaultProps} />);

            expect(accordion).toMatchSnapshot();
        });

        it("expands a group", () => {
            const accordion = mount(<Accordion {...defaultProps} />);

            accordion.find(".widget-accordion-group-header").simulate("click");
            expect(accordion).toMatchSnapshot();
        });

        it("allows one group to be expanded only", () => {
            const groups = [...defaultProps.groups];
            groups[1].visible = true;
            const accordion = mount(<Accordion {...defaultProps} groups={groups} />);

            accordion.find(".widget-accordion-group-header").first().simulate("click");
            accordion.find(".widget-accordion-group-header").last().simulate("click");
            expect(accordion).toMatchSnapshot();
        });

        it("collapses a group", () => {
            const accordion = mount(<Accordion {...defaultProps} />);

            accordion.find(".widget-accordion-group-header").first().simulate("click");
            accordion.find(".widget-accordion-group-header").first().simulate("click");
            expect(accordion).toMatchSnapshot();
        });
    });

    describe("in collapsible & multiple expanded group mode", () => {
        beforeEach(() => {
            defaultProps = getProps(true, false);
        });

        it("renders correctly", () => {
            const accordion = mount(<Accordion {...defaultProps} />);

            expect(accordion).toMatchSnapshot();
        });

        it("expands a group", () => {
            const accordion = mount(<Accordion {...defaultProps} />);

            accordion.find(".widget-accordion-group-header").simulate("click");
            expect(accordion).toMatchSnapshot();
        });

        it("allows multiple groups to be expanded", () => {
            const groups = [...defaultProps.groups];
            groups[1].visible = true;
            const accordion = mount(<Accordion {...defaultProps} groups={groups} />);

            accordion.find(".widget-accordion-group-header").first().simulate("click");
            accordion.find(".widget-accordion-group-header").last().simulate("click");
            expect(accordion).toMatchSnapshot();
        });

        it("collapses a group", () => {
            const groups = [...defaultProps.groups];
            groups[1].visible = true;
            const accordion = mount(<Accordion {...defaultProps} groups={groups} />);

            accordion.find(".widget-accordion-group-header").first().simulate("click");
            accordion.find(".widget-accordion-group-header").last().simulate("click");
            accordion.find(".widget-accordion-group-header").first().simulate("click");
            expect(accordion).toMatchSnapshot();
        });
    });

    describe("not collapsible", () => {
        beforeEach(() => {
            defaultProps = getProps(false);
        });

        it("renders correctly", () => {
            const accordion = mount(<Accordion {...defaultProps} />);

            expect(accordion).toMatchSnapshot();
        });
    });
});
