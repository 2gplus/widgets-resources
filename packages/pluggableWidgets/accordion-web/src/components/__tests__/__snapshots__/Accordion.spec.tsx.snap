// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion in collapsible & multiple expanded group mode renders correctly 1`] = `
<ContextProvider
  value={[Function]}
>
  <div
    className=""
    id="id"
    style={
      Object {
        "height": "500px",
      }
    }
    tabIndex={1}
  >
    <AccordionGroup
      group={
        Object {
          "collapsed": true,
          "content": <span>
            content
          </span>,
          "header": "header",
          "visible": true,
        }
      }
      key="0"
    />
    <AccordionGroup
      group={
        Object {
          "collapsed": true,
          "content": <span>
            content2
          </span>,
          "header": "header2",
          "visible": false,
        }
      }
      key="1"
    />
  </div>
</ContextProvider>
`;

exports[`Accordion in collapsible & multiple expanded group mode updates when accordion group data changes 1`] = `
<ContextProvider
  value={[Function]}
>
  <div
    className=""
    id="id"
    style={
      Object {
        "height": "500px",
      }
    }
    tabIndex={1}
  >
    <AccordionGroup
      group={
        Object {
          "collapsed": true,
          "content": <span>
            content
          </span>,
          "header": "header",
          "visible": true,
        }
      }
      key="0"
    />
    <AccordionGroup
      group={
        Object {
          "collapsed": true,
          "content": <span>
            content2
          </span>,
          "header": "header2",
          "visible": true,
        }
      }
      key="1"
    />
  </div>
</ContextProvider>
`;

exports[`Accordion in collapsible & single expanded group mode renders correctly 1`] = `
<ContextProvider
  value={[Function]}
>
  <div
    className=""
    id="id"
    style={
      Object {
        "height": "500px",
      }
    }
    tabIndex={1}
  >
    <AccordionGroup
      group={
        Object {
          "collapsed": true,
          "content": <span>
            content
          </span>,
          "header": "header",
          "visible": true,
        }
      }
      key="0"
    />
    <AccordionGroup
      group={
        Object {
          "collapsed": true,
          "content": <span>
            content2
          </span>,
          "header": "header2",
          "visible": false,
        }
      }
      key="1"
    />
  </div>
</ContextProvider>
`;

exports[`Accordion in collapsible & single expanded group mode updates when accordion group data changes 1`] = `
<ContextProvider
  value={[Function]}
>
  <div
    className=""
    id="id"
    style={
      Object {
        "height": "500px",
      }
    }
    tabIndex={1}
  >
    <AccordionGroup
      group={
        Object {
          "collapsed": true,
          "content": <span>
            content
          </span>,
          "header": "header",
          "visible": true,
        }
      }
      key="0"
    />
    <AccordionGroup
      group={
        Object {
          "collapsed": true,
          "content": <span>
            content2
          </span>,
          "header": "header2",
          "visible": true,
        }
      }
      key="1"
    />
  </div>
</ContextProvider>
`;

exports[`Accordion not collapsible renders correctly 1`] = `
<ContextProvider>
  <div
    className=""
    id="id"
    style={
      Object {
        "height": "500px",
      }
    }
    tabIndex={1}
  >
    <AccordionGroup
      group={
        Object {
          "collapsed": false,
          "content": <span>
            content
          </span>,
          "header": "header",
          "visible": true,
        }
      }
      key="0"
    />
    <AccordionGroup
      group={
        Object {
          "collapsed": false,
          "content": <span>
            content2
          </span>,
          "header": "header2",
          "visible": false,
        }
      }
      key="1"
    />
  </div>
</ContextProvider>
`;

exports[`Accordion not collapsible updates when accordion group data changes 1`] = `
<ContextProvider>
  <div
    className=""
    id="id"
    style={
      Object {
        "height": "500px",
      }
    }
    tabIndex={1}
  >
    <AccordionGroup
      group={
        Object {
          "collapsed": false,
          "content": <span>
            content
          </span>,
          "header": "header",
          "visible": true,
        }
      }
      key="0"
    />
    <AccordionGroup
      group={
        Object {
          "collapsed": false,
          "content": <span>
            content2
          </span>,
          "header": "header2",
          "visible": true,
        }
      }
      key="1"
    />
  </div>
</ContextProvider>
`;
